sdk/appconfiguration/azure-appconfiguration-provider/azure/appconfiguration/provider/_client_manager_base.py

```diff
@@ -25,24 +26,104 @@
     ETAG_KEY,
     FEATURE_FLAG_REFERENCE_KEY,
     FEATURE_FLAG_ID_KEY,
+    ALLOCATION_ID_KEY,
 )
 
 FALLBACK_CLIENT_REFRESH_EXPIRED_INTERVAL = 3600  # 1 hour in seconds
 MINIMAL_CLIENT_REFRESH_INTERVAL = 30  # 30 seconds
 
+JSON = Mapping[str, Any]
+
 
 @dataclass
 class _ConfigurationClientWrapperBase:
     endpoint: str
 
+    @staticmethod
+    def _generate_allocation_id(feature_flag_value: Dict[str, JSON]) -> Optional[str]:
+        """
+        Generates an allocation ID for the specified feature.
+        seed=123abc\ndefault_when_enabled=Control\npercentiles=0,Control,20;20,Test,100\nvariants=Control,standard;Test,special # pylint:disable=line-too-long
+
+        :param  Dict[str, JSON] feature_flag_value: The feature to generate an allocation ID for.
+        :rtype: str
+        :return: The allocation ID.
+        """
+
+        allocation_id = ""
+        allocated_variants = []
+
+        allocation: Optional[JSON] = feature_flag_value.get("allocation")
+
+        if allocation:
+            # Seed
+            allocation_id = f"seed={allocation.get('seed', '')}"
+
+            # DefaultWhenEnabled
+            if "default_when_enabled" in allocation:
+                allocated_variants.append(allocation.get("default_when_enabled"))
+
+            allocation_id += f"\ndefault_when_enabled={allocation.get('default_when_enabled', '')}"
+
+            # Percentile
+            allocation_id += "\npercentiles="
+
+            percentile = allocation.get("percentile")
+
+            if percentile:
+                percentile_allocations = sorted(
+                    (x for x in percentile if x.get("from") != x.get("to")),
```

Comment:
Should we validate that from, to and variant fields are all not None?

Reply:
From logic in the creation of percentile allocation to must be an int so it will raise an error on creation if it isn't. from is optional[int], so we have to check it.